cmd/umoci/copy-remote-to-local.go

```diff
@@ -0,0 +1,217 @@
+	godigest "github.com/opencontainers/go-digest"
+	v1 "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/opencontainers/umoci/oci/cas/dir"
+	"github.com/opencontainers/umoci/oci/casext"
+	"github.com/pkg/errors"
+	"github.com/urfave/cli"
+)
+
+// a.k.a. "pull"
+func copyRemoteToLocal(ctx *cli.Context, remote *parsedRemoteReference, local *parsedLocalReference) error {
+	client, err := newRegistryClient(ctx, remote)
+	if err != nil {
+		return err
+	}
+
+	// Get a reference to the CAS.
+	dir.Create(local.dir) // TODO: handle this error?
+	engine, err := dir.Open(local.dir)
+	if err != nil {
+		return errors.Wrap(err, "open CAS")
+	}
+	engineExt := casext.NewEngine(engine)
+	defer engine.Close()
+	engineContext := context.Background()
+
+	log.Infof("Checking if manifest available in registry")
+	req := client.NewRequest(reggie.HEAD, "/v2/<name>/manifests/<reference>",
+		reggie.WithReference(remote.tag)).
+		SetHeader("Accept", v1.MediaTypeImageManifest)
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+	expectedContentDigest := resp.Header().Get(contentDigestHeader)
+	log.Infof("Registry reports manifest with digest %s", expectedContentDigest)
+	parsedDigest, err := godigest.Parse(expectedContentDigest)
+	if err != nil {
+		return err
+	}
+
+	// download manifest if it doesnt already exist is local store
+	var manifestBytes []byte
+	manifestReader, err := engine.GetBlob(engineContext, parsedDigest)
+	defer manifestReader.Close()
+	if err != nil {
+		// TODO: better than this error check
+		if !strings.Contains(err.Error(), "no such file") {
+			return err
+		}
+		// Fetch the raw manifest from registry and validate its digest
+		log.Infof("Downloading manifest from registry")
+		req = client.NewRequest(reggie.GET, "/v2/<name>/manifests/<reference>",
+			reggie.WithReference(remote.tag)).
+			SetHeader("Accept", v1.MediaTypeImageManifest)
+		resp, err = client.Do(req)
+		if err != nil {
+			return err
+		}
+		if h := resp.Header().Get(contentDigestHeader); h != expectedContentDigest {
+			return errors.New(
+				fmt.Sprintf("Possible MITM attack: the %s header was %s on manifest HEAD, but %s on manifest GET",
+					contentDigestHeader, expectedContentDigest, h))
+		}
+		actualContentDigest := godigest.FromBytes(resp.Body()).String()
+		if actualContentDigest != expectedContentDigest {
+			return errors.New(
+				fmt.Sprintf("Possible MITM attack: the real digest of the downloaded manifest was %s",
+					actualContentDigest))
+		}
+		log.Debugf("actual manifest digest matches expected digest (%s)", expectedContentDigest)
+
+		// Note: only "application/vnd.oci.image.manifest.v1+json" supported for now
+		mediaType := resp.Header().Get(manifestMediaTypeHeader)
+		if mediaType != v1.MediaTypeImageManifest {
+			return errors.New(
+				fmt.Sprintf("Content-Type header for image manifest invalid: %s", mediaType))
+		}
+
+		d, s, err := engine.PutBlob(engineContext, bytes.NewReader(resp.Body()))
```

Comment:
As far as I can tell, resp.Body() here is: https://pkg.go.dev/github.com/go-resty/resty/v2#Response.Body which will end up reading the whole layer into memory. Seems like you want RawBody() here?

https://pkg.go.dev/github.com/go-resty/resty/v2#Response.RawBody

Reply:
Oh, derp. I meant to put this on the layer one below. For this one I think it's unavoidable: you definitely want to read the manifest bytes into memory.